### PR TITLE
🎉 New Source: Yotpo [Low code CDK]

### PR DESCRIPTION
**Pull request by btkcodedev**
## What
*Developing new connector for the source: Yotpo*

Resolves:
[Yotpo](https://github.com/airbytehq/DeveloperWeek-Connector-Contest/issues/14)

## How
*Developed using (Configuration Based Source) low-code CDK*

## Recommended reading order
1. `spec.yaml`
2. `manifest.yaml`
3. `schemas/*`

## Tests

<details open><summary><strong>Integration & Acceptance</strong></summary>
Full Test Results

image

</details>

## 🚨 User Impact 🚨
- No breaking changes, Just addition of new source
<details><summary><strong>New Connector: Source Yotpo</strong></summary>

### Community member or Airbyter

- [x] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [x] Code reviews completed
- [ ] Documentation updated
    - [x] Connector's `README.md`
    - [x] Connector's `bootstrap.md`
    - [x] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [x] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

</details>

### CODE REVIEW

**Comment by btkcodedev** [comment]
@RealChrisSean 
All acceptance tests are passing in local (Attached screenshot as description)
Please assign it for review :)

**Comment by btkcodedev** [comment]
PS for reviewers: 
Please ping via slack for getting sandbox account credentials and secrets for testing

**Comment by marcosmarxm** [comment]
Thanks for the contribution @btkcodedev right now our team is reviewing GA connector changes and can take a while to get to your contribution. Thanks for your patience.

File: airbyte-integrations/connectors/source-yotpo/source_yotpo/manifest.yaml [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    type: DeclarativeStream
+    retriever:
+      type: SimpleRetriever
+      record_selector:
+        type: RecordSelector
+        extractor:
+          type: DpathExtractor
+          field_path: ["reviews"]
+      paginator:
+        type: NoPagination
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-yotpo/source_yotpo/manifest.yaml:56]: Please add more data into the stream to validate the paginator.

File: airbyte-integrations/connectors/source-yotpo/source_yotpo/manifest.yaml [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+      requester:
+        type: HttpRequester
+        url_base: "https://api.yotpo.com"
+        http_method: "GET"
+        request_parameters:
+          utoken: "{{ config['access_token'] }}"
+        authenticator:
+          type: BearerAuthenticator
+          api_token: "{{ config['access_token'] }}"
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-yotpo/source_yotpo/manifest.yaml:103]: Create a YoptoRequester to reuse for all streams.

File: airbyte-integrations/connectors/source-yotpo/source_yotpo/schemas/webhooks.json [hunk 3]
```diff
@@ -0,0 +1,30 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "Webhooks Schema",
+  "type": "object",
+  "properties": {
+    "webhooks": {
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-yotpo/source_yotpo/schemas/webhooks.json:6]: Please retrieve the nested item not the array.

File: airbyte-integrations/connectors/source-yotpo/source_yotpo/manifest.yaml [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+  webhooks_stream:
+    type: DeclarativeStream
+    retriever:
+      type: SimpleRetriever
+      record_selector:
+        type: RecordSelector
+        extractor:
+          type: DpathExtractor
+          field_path: ["response"]
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-yotpo/source_yotpo/manifest.yaml:141]:
```suggestion
          field_path: ["response", "webhooks"]
```

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
Left some comments please check them @btkcodedev

File: airbyte-integrations/connectors/source-yotpo/source_yotpo/manifest.yaml (hunk 2 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-yotpo/source_yotpo/manifest.yaml]: Actually 3 streams differ from each other, top_reviews only results 5 star rating, reviews has incremental reads and other one reads review with no ratings, Streams will be useful for the users

File: airbyte-integrations/connectors/source-yotpo/source_yotpo/schemas/reviews.json [hunk 5]
```diff
@@ -0,0 +1,60 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "Reviews Schema",
+  "type": "object",
```
[the remark this answers is not in the document]
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-yotpo/source_yotpo/schemas/reviews.json:5]: Done

File: airbyte-integrations/connectors/source-yotpo/source_yotpo/manifest.yaml [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    $parameters:
+      path: "/apps/{{ config['app_key'] }}/webhooks"
+
+streams:
+  - "#/definitions/privacy_data_stream"
+  - "#/definitions/privacy_user_stream"
+  - "#/definitions/reviews_stream"
+  - "#/definitions/reviews_bottomline_stream"
+  - "#/definitions/top_reviews_stream"
+  - "#/definitions/webhooks_stream"
```
[the remark this answers is not in the document]
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-yotpo/source_yotpo/manifest.yaml:132]: Product endpoint feature is disabled in free plan
Unsubscribers is empty stream and post method rehires paid plan

Will add other streams

**Comment by btkcodedev** [comment]
All comments resolved :)

**Comment by btkcodedev** [comment]
Test passed

image

File: airbyte-integrations/connectors/source-yotpo/source_yotpo/manifest.yaml [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+          field_path: ["reviews"]
+      paginator:
+        type: "DefaultPaginator"
+        page_size_option:
+          type: "RequestOption"
+          inject_into: "request_parameter"
+          field_name: "per_page"
+        pagination_strategy:
+          type: "PageIncrement"
+          page_size: 5
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-yotpo/source_yotpo/manifest.yaml:57]: Where did you find this information in Yopto docs?

File: airbyte-integrations/connectors/source-yotpo/source_yotpo/manifest.yaml [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+      utoken: "{{ config['access_token'] }}"
+
+  reviews_stream:
+    type: DeclarativeStream
+    retriever:
+      type: SimpleRetriever
+      record_selector:
+        type: RecordSelector
+        extractor:
+          type: DpathExtractor
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-yotpo/source_yotpo/manifest.yaml:47]: Please create a base_stream component to not repeat so many times this. Take a look in other connectors how to do it.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-yotpo/source_yotpo/manifest.yaml (hunk 2 above)
Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-yotpo/source_yotpo/manifest.yaml]: But top_reviews and reviews bottomline aren't only a subset of reviews?

File: airbyte-integrations/connectors/source-yotpo/source_yotpo/manifest.yaml (hunk 1 above)
Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-yotpo/source_yotpo/manifest.yaml:109]: @btkcodedev don't resolve conversation if they weren't solved.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-yotpo/source_yotpo/manifest.yaml (hunk 2 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-yotpo/source_yotpo/manifest.yaml]: That is correct :) Removed.

File: airbyte-integrations/connectors/source-yotpo/source_yotpo/manifest.yaml (hunk 7 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-yotpo/source_yotpo/manifest.yaml]: It was incorrect, Corrected :)

**Comment by btkcodedev** [comment]
Well well, @marcosmarxm, I've checked the paginator as per https://apidocs.yotpo.com/reference/guidelines, and it was working, But it breaks the incremental read with the following error
image

Thus rolled back to only one review and thus connector passed the acceptance tests

**Comment by btkcodedev** [comment]
image

Test results

File: airbyte-integrations/connectors/source-yotpo/source_yotpo/manifest.yaml (hunk 7 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-yotpo/source_yotpo/manifest.yaml]: Resolved

File: airbyte-integrations/connectors/source-yotpo/source_yotpo/manifest.yaml (hunk 8 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-yotpo/source_yotpo/manifest.yaml]: Done

**Comment by btkcodedev** [comment]
@marcosmarxm Little Bump Here, cause the deadline is by 27th🙂

**Comment by btkcodedev** [comment]
@marcosmarxm Pushing from https://github.com/airbytehq/airbyte/pull/25532

**Pull request closed without merging**